lib/grammers-client/src/types/chats.rs

```diff
@@ -27,54 +31,134 @@ pub struct AdminRightsBuilder {
+        } else {
+            Err(InvocationError::Rpc(RpcError {
+                code: 400,
+                name: "PEER_ID_INVALID".to_string(),
+                value: None,
+            }))
+        }
+    }
+}
+
+pin_project! {
+    /// Builder for editing the administrator rights of a user in a specific chat.
+    ///
+    /// Use [`Client::set_admin_rights`] to retrieve an instance of this type.
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub struct AdminRightsBuilder<F: Future<Output = BuilderRes>> {
+        inner: Option<AdminRightsBuilderInner>,
+        gen: AdminFutGen<F>,
+        #[pin]
+        fut: Option<F>
+    }
+}
+
+impl<F: Future<Output = BuilderRes>> Future for AdminRightsBuilder<F> {
+    type Output = BuilderRes;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<BuilderRes> {
+        let mut s = self.project();
+        if s.fut.is_none() {
+            // unwrap safety: s.inner is None only when s.fut is some
+            // or s.fut is resolved
+            s.fut.set(Some((s.gen)(s.inner.take().unwrap())))
+        }
+
+        s.fut.as_pin_mut().unwrap().poll(cx)
+    }
+}
+
+impl<F: Future<Output = BuilderRes>> AdminRightsBuilder<F> {
+    pub(crate) fn new(client: Client, chat: &Chat, user: &User, gen: AdminFutGen<F>) -> Self {
         Self {
-            client,
-            chat: chat.clone(),
-            peer: user.to_input_peer(),
-            user: user.to_input(),
-            rank: "".into(),
-            rights: tl::types::ChatAdminRights {
-                anonymous: false,
-                change_info: false,
-                post_messages: false,
-                edit_messages: false,
-                delete_messages: false,
-                ban_users: false,
-                invite_users: false,
-                pin_messages: false,
-                add_admins: false,
-                manage_call: false,
-                other: false,
-            },
+            inner: Some(AdminRightsBuilderInner {
+                client,
+                chat: chat.clone(),
+                peer: user.to_input_peer(),
+                user: user.to_input(),
+                rank: "".into(),
+                rights: tl::types::ChatAdminRights {
+                    anonymous: false,
+                    change_info: false,
+                    post_messages: false,
+                    edit_messages: false,
+                    delete_messages: false,
+                    ban_users: false,
+                    invite_users: false,
+                    pin_messages: false,
+                    add_admins: false,
+                    manage_call: false,
+                    other: false,
+                },
+            }),
+            gen,
+            fut: None,
         }
     }
 
+    fn inner_mut<'a>(&'a mut self) -> &'a mut AdminRightsBuilderInner {
+        // Unwrap safety: AdminRightsBuilderInner should never be None unless polled after being
```

Comment:
Does the "safety" here mean it would be `unsafe` to `panic!`?

Reply:
> Does the "safety" here mean it would be `unsafe` to `panic!`?

No it means inner_mut will never panic.
`.await` will consume the builder. (i.e self.inner.poll is called)
`self.inner` only becomes `None` after being `poll`ed.